File: simulator/backend/services/competency-service.ts
/**
 * Competency Service
 * Business logic for competency operations
 */

import { competencyRepository } from '../db/repositories/file-competency-repository.ts';
import type { CompetencyData } from '@/shared/types/api.types';

export class CompetencyService {
  /**
   * Get competency by ID
   */
  async getCompetencyById(id: string): Promise<CompetencyData | null> {
    try {
      return await competencyRepository.findById(id);
    } catch (error) {
      console.error('Error getting competency by ID:', error);
      throw error;
    }
  }

  /**
   * Get all competencies
   */
  async getAllCompetencies(): Promise<CompetencyData[]> {
    try {
      return await competencyRepository.findAll();
    } catch (error) {
      console.error('Error getting all competencies:', error);
      throw error;
    }
  }

  /**
   * Get competencies by industry
   */
  async getCompetenciesByIndustry(industry: string): Promise<CompetencyData[]> {
    try {
      return await competencyRepository.findByIndustry(industry);
    } catch (error) {
      console.error('Error getting competencies by industry:', error);
      throw error;
    }
  }

  /**
   * Get competencies by category
   */
  async getCompetenciesByCategory(category: string): Promise<CompetencyData[]> {
    try {
      return await competencyRepository.findByCategory(category);
    } catch (error) {
      console.error('Error getting competencies by category:', error);
      throw error;
    }
  }

  /**
   * Create competency
   */
  async createCompetency(competencyData: Omit<CompetencyData, 'id'>): Promise<CompetencyData> {
    try {
      // Validate competency data
      this.validateCompetency(competencyData);

      return await competencyRepository.create(competencyData);
    } catch (error) {
      console.error('Error creating competency:', error);
      throw error;
    }
  }

  /**
   * Update competency
   */
  async updateCompetency(id: string, competencyData: Partial<CompetencyData>): Promise<CompetencyData> {
    try {
      const existing = await competencyRepository.findById(id);
      if (!existing) {
        throw new Error('Competency not found');
      }

      return await competencyRepository.update(id, competencyData);
    } catch (error) {
      console.error('Error updating competency:', error);
      throw error;
    }
  }

  /**
   * Delete competency
   */
  async deleteCompetency(id: string): Promise<void> {
    try {
      const existing = await competencyRepository.findById(id);
      if (!existing) {
        throw new Error('Competency not found');
      }

      await competencyRepository.delete(id);
    } catch (error) {
      console.error('Error deleting competency:', error);
      throw error;
    }
  }

  /**
   * Get all categories
   */
  async getAllCategories(): Promise<string[]> {
    try {
      return await competencyRepository.getAllCategories();
    } catch (error) {
      console.error('Error getting competency categories:', error);
      throw error;
    }
  }

  /**
   * Get competencies grouped by category
   */
  async getCompetenciesGroupedByCategory(): Promise<Record<string, CompetencyData[]>> {
    try {
      const competencies = await competencyRepository.findAll();
      const grouped: Record<string, CompetencyData[]> = {};

      competencies.forEach(competency => {
        if (!grouped[competency.category]) {
          grouped[competency.category] = [];
        }
        grouped[competency.category].push(competency);
      });

      return grouped;
    } catch (error) {
      console.error('Error getting competencies grouped by category:', error);
      throw error;
    }
  }

  /**
   * Validate competency data
   */
  private validateCompetency(competencyData: Partial<CompetencyData>): void {
    if (!competencyData.name) {
      throw new Error('Competency name is required');
    }

    if (!competencyData.category) {
      throw new Error('Competency category is required');
    }

    if (competencyData.weight !== undefined) {
      if (competencyData.weight < 0 || competencyData.weight > 100) {
        throw new Error('Competency weight must be between 0 and 100');
      }
    }
  }

  /**
   * Calculate weighted score
   */
  calculateWeightedScore(scores: Record<string, number>, competencies: CompetencyData[]): number {
    try {
      let totalWeight = 0;
      let weightedSum = 0;

      competencies.forEach(competency => {
        const score = scores[competency.id] || 0;
        const weight = competency.weight || 1;

        weightedSum += score * weight;
        totalWeight += weight;
      });

      return totalWeight > 0 ? weightedSum / totalWeight : 0;
    } catch (error) {
      console.error('Error calculating weighted score:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const competencyService = new CompetencyService();

File: backend-python/app/services/competency_service.py
"""
Competency Service
Business logic for competency operations
"""

import logging
from typing import Any

from app.repositories.competency_repository import competency_repository
from app.repositories.file_competency_repository import file_competency_repository
from app.models.competency import CompetencyData, CreateCompetencyRequest, UpdateCompetencyRequest

logger = logging.getLogger(__name__)


class CompetencyService:
    """Service for competency operations."""

    def __init__(self, use_file_repository: bool = True):
        """Initialize with either file or database repository."""
        self.repository = file_competency_repository if use_file_repository else competency_repository

    async def get_competency_by_id(self, competency_id: str) -> CompetencyData | None:
        """Get competency by ID."""
        try:
            return await self.repository.find_by_id(competency_id)
        except Exception as e:
            logger.error(f"Error getting competency by ID: {e}")
            raise

    async def get_all_competencies(self) -> list[CompetencyData]:
        """Get all competencies."""
        try:
            return await self.repository.find_all()
        except Exception as e:
            logger.error(f"Error getting all competencies: {e}")
            raise

    async def get_competencies_by_industry(self, industry: str) -> list[CompetencyData]:
        """Get competencies by industry."""
        try:
            return await self.repository.find_by_industry(industry)
        except Exception as e:
            logger.error(f"Error getting competencies by industry: {e}")
            raise

    async def get_competencies_by_category(self, category: str) -> list[CompetencyData]:
        """Get competencies by category."""
        try:
            return await self.repository.find_by_category(category)
        except Exception as e:
            logger.error(f"Error getting competencies by category: {e}")
            raise

    async def create_competency(self, competency_data: CreateCompetencyRequest | dict[str, Any]) -> CompetencyData:
        """Create new competency."""
        try:
            return await self.repository.create(competency_data)
        except Exception as e:
            logger.error(f"Error creating competency: {e}")
            raise

    async def update_competency(
        self,
        competency_id: str,
        competency_data: UpdateCompetencyRequest | dict[str, Any],
    ) -> CompetencyData:
        """Update competency."""
        try:
            return await self.repository.update(competency_id, competency_data)
        except Exception as e:
            logger.error(f"Error updating competency: {e}")
            raise

    async def delete_competency(self, competency_id: str) -> None:
        """Delete competency."""
        try:
            await self.repository.delete(competency_id)
        except Exception as e:
            logger.error(f"Error deleting competency: {e}")
            raise

    async def get_categories(self) -> list[str]:
        """Get all unique categories."""
        try:
            competencies = await self.repository.find_all()
            categories = set()
            for comp in competencies:
                if comp.category:
                    categories.add(comp.category)
            return sorted(list(categories))
        except Exception as e:
            logger.error(f"Error getting categories: {e}")
            raise


# Singleton instance
competency_service = CompetencyService(use_file_repository=True)
